### PR TITLE
Sw 30/battery charge

### PR DESCRIPTION
**Pull request by jrvollmer**
Several changes/additions:
       - Graphs
       - Battery charge component
       - TCP communication with solar car
       - README update

### CODE REVIEW

File: Backend/src/routes/api.js [hunk 1]
```diff
@@ -17,24 +17,29 @@ export default router;
 //--------------TCP-------------------------
 
 import { Socket } from "net";
-// const { Buffer } = require("buffer");
-const car_port = 4003;
+const car_port = 4003; // Port for TCP connection
```
Comment (EUdds) [inline Backend/src/routes/api.js]: Constant variables should be all caps


```suggestion
const CAR_PORT = 4003; // Port for TCP connection
```

Honestly they should be kept in some one singular file but let's discuss later.

File: Backend/src/routes/api.js [hunk 2]
```diff
@@ -47,32 +52,55 @@ client.on("error", (err) => {
...
-        );
+        // Add the data to the front of dataArray
+        dataArray.unshift(String.fromCharCode(data.readUInt8(buffOffset)));
         break;
       case "bool":
-        solarCarData[property] = Boolean(data.readUInt8(buffOffset));
+        // Add the data to the front of dataArray
+        dataArray.unshift(Boolean(data.readUInt8(buffOffset)));
         break;
       default:
```
Comment (EUdds) [inline Backend/src/routes/api.js:93]: Should we add something that happens here?

File: DataGenerator/src/server.js [hunk 3]
```diff
@@ -51,25 +38,27 @@ const server = net.createServer((socket) => {
...
-          buf1.writeFloatBE(nextValue + 0.125, buffOffset);
+          buf1.writeFloatLE(Math.floor(nextValue) + 0.125, buffOffset);
           break;
         case "char":
-          buf1.writeUInt8(nextValue, buffOffset);
+          buf1.writeUInt8(Math.round(nextValue), buffOffset);
           break;
         case "bool":
-          buf1.writeUInt8(nextValue % 2, buffOffset);
+          buf1.writeUInt8(Math.round(nextValue) % 2, buffOffset);
```
Comment (EUdds) [inline DataGenerator/src/server.js:63]: Why % 2? Can we put a comment or put it in a function?

File: Frontend/src/Components/Graph/CustomGraph.js [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+        setDatasetButtons(datasetButtons);
+    };
+
+    // ------------------------- Change/Validate title ---------------------------
+
+    const changeTitle = (event) => {
+        setTitle(event.target.value);
+    };
+
+    const isInvalidTitle = ((title === "") || (title === "custom") || (title === "Custom"));
```
Comment (EUdds) [inline Frontend/src/Components/Graph/CustomGraph.js]:
```suggestion
    const isInvalidTitle = ((title === "") || (title.toLowerCase() === "custom"));
```

**Review by EUdds — CHANGES_REQUESTED** [review]
Few suggestions

File: Backend/src/routes/api.js (hunk 2 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js:93]: Could you please elaborate? Is this in reference to the default case?

Comment (EUdds) [reply Backend/src/routes/api.js:93]: Yeah right now nothing happens, should we log or warn or something here?

Comment (jrvollmer) [reply Backend/src/routes/api.js:93]: Yeah. I added a log for unexpected data types

File: DataGenerator/src/server.js (hunk 3 above)
Comment (jrvollmer) [reply DataGenerator/src/server.js:63]: Added a comment

File: Frontend/src/Components/Graph/CustomGraph.js (hunk 4 above)
Comment (jrvollmer) [reply Frontend/src/Components/Graph/CustomGraph.js]: Yeah, allowing CuSToM was a bit odd. Fixed

File: Backend/src/routes/api.js (hunk 1 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: I changed constants to all caps, and I made a note/task to add a constants file.

**Review by EUdds — APPROVED** [review]

**Pull request merged**